lib/wit/dependency.py

```diff
@@ -0,0 +1,154 @@
+        for subdep in subdeps:
+            subdep.load_package(packages, repo_paths)
+            subdep.package.load_repo(wsroot, download, subdep.specified_revision)
+
+            if subdep.name in source_map:
+                if subdep.package.resolve_source(subdep.source) != source_map[subdep.name]:
+                    log.error("Dependency [{}] has multiple conflicting paths:\n"
+                              "  {}\n"
+                              "  {}\n".format(subdep.name, subdep.source,
+                                              source_map[subdep.name]))
+                    sys.exit(1)
+
+            source_map[subdep.name] = subdep.package.resolve_source(subdep.source)
+
+            if subdep.package.repo is None:
+                continue
+
+            if subdep.get_commit_time() > self.get_commit_time():
+                log.error("Repo [{}] has a dependent that is newer than the source. "
+                          "This should not happen.\n".format(subdep.name))
+                sys.exit(1)
+
+            commit_time = subdep.get_commit_time()
+            queue.append((commit_time, subdep))
+
+        queue.sort(key=lambda tup: tup[0])
+
+        return source_map, packages, queue
+
+    def __key(self):
+        return (self.source, self.specified_revision, self.name)
+
+    def __hash__(self):
+        return hash(self.__key())
+
+    def __eq__(self, other):
+        return isinstance(self, type(other)) and self.__key() == other.__key()
+
+    @staticmethod
+    def infer_name(source):
+        return Path(source).name.replace('.git', '')
+
+    # wsroot, repo_paths, packages, force_root
+    # NB: mutates packages[self.name]!
+    def load_package(self, packages, repo_paths):
+        """
+        Bind itself to a package, using one in the `packages` argument if available.
+        This will also add itself as a dependent of the package
+        """
+        if self.name in packages:
+            self.package = packages[self.name]
+        else:
+            self.package = Package(self.name, repo_paths)
+            self.package.set_source(self.source)
+            packages[self.name] = self.package
+        self.package.add_dependent(self)
+
+    def add_dependent(self, dependent):
+        if dependent not in self.dependents:
+            self.dependents.append(dependent)
+
+    def get_commit_time(self):
+        return datetime.utcfromtimestamp(int(self.package.repo.commit_to_time(
+            self.specified_revision)))
+
+    def manifest(self):
+        return {
+            'name': self.name,
+            'source': self.source,
+            'commit': self.specified_revision,
+        }
+
+    # used before saving to manifests/lockfiles
+    def resolved(self):
+        if self.package is None or self.package.repo is None:
+            raise Exception("Cannot resolve dependency that us unbound to disk")
+        return Dependency(self.name, self.source, self.resolved_rev())
```

Comment:
```suggestion
    def resolved(self):
        rev = self.resolved_rev()
        return Dependency(self.name, self.source, rev)
```

Reply:
The diff shows up wrong, but the idea being that the error check is redundant with that in `resolved_rev`.

Reply:
Good catch. That diff did look weird so I fixed it with https://github.com/sifive/wit/pull/96/commits/f27443d7754a8c326cd8afb4f01ebbace2b4961a